src/Oscoin/Protocol/Sync/RealWorld.hs

```diff
@@ -1,84 +1,136 @@
-    ( Serialise (ProtocolResponse c tx s 'GetTip) )
+    ( Serialise tx
+    , Serialise s
+    , Serialise (Hash c)
+    , Serialise (PublicKey c)
+    , HasHashing c
+    , Serialise (ProtocolResponse c ResIO tx s 'GetTip)
+    , ProtocolResponse c ResIO tx s 'GetBlockHeaders ~ OldestFirst [] (BlockHeader c (Sealed c s))
+    , ProtocolResponse c ResIO tx s 'GetBlocks ~ ConduitT () (Block c tx (Sealed c s)) ResIO ()
+    )
     => HTTP.Manager
-    -> DataFetcher c tx s IO
-newDataFetcherIO httpManager = DataFetcher $ \peer -> \case
-    -- This implementation is basically 'fetchTip' function from the spec.
-    SGetTip -> \() -> timed maxAllowedNetworkLatency $ do
-        let hn = addrHost . nodeHttpApiAddr $ peer
+    -> DataFetcher c tx s ResIO
+newDataFetcher httpManager = DataFetcher $ \peer args ->
+    let prettyHost = renderHost (addrHost . nodeHttpApiAddr $ peer)
         -- FIXME(adn) Support for https if needed.
-        rq0 <- HTTP.parseUrlThrow ("http://" <> toS (renderHost hn) <> "/blockchain/tip")
-        let rq = rq0 { HTTP.port = fromIntegral (addrPort . nodeHttpApiAddr $ peer) }
-
-       -- Trying to issue a wrong response type will trigger a compilation
-       -- error similar to:
-       -- • Couldn't match type ‘'GetBlockHeaders’ with ‘'GetTip’
-       -- Expected type: IO (ProtocolResponse r)
-       -- Actual type: IO (ProtocolResponse 'GetBlockHeaders)
-
-        HTTP.withResponse rq httpManager $ \res -> do
-            cborBlob <- HTTP.brRead (HTTP.responseBody res)
-            pure $ CBOR.deserialise . toS $ cborBlob
-
-    -- This implementation is basically 'fetchBlocks' from the spec.
-    SGetBlocks -> \Range{..} -> notImplemented
-
-    -- This implementation is basically 'fetchBlockHeaders' from the spec.
-    SGetBlockHeaders -> \Range{..} -> notImplemented
+        protocol   = "http://"
+        prefix     = protocol <> prettyHost
+    in case args of
+        -- This implementation is basically 'fetchTip' function from the spec.
+        SGetTip -> \() -> timed maxAllowedNetworkLatency $ do
+            let url = prefix <> "/blockchain/tip"
+            rq0 <- HTTP.parseUrlThrow (toS url)
+            let rq = rq0 { HTTP.port = fromIntegral (addrPort . nodeHttpApiAddr $ peer) }
+
+           -- Trying to issue a wrong response type will trigger a compilation
+           -- error similar to:
+           -- • Couldn't match type ‘'GetBlockHeaders’ with ‘'GetTip’
+           -- Expected type: IO (ProtocolResponse r)
+           -- Actual type: IO (ProtocolResponse 'GetBlockHeaders)
+
+            liftIO $ HTTP.withResponse rq httpManager $ \res -> do
+                cborBlob <- HTTP.brRead (HTTP.responseBody res)
+                pure $ CBOR.deserialise . toS $ cborBlob
+
+        -- This implementation is basically 'fetchBlocks' from the spec.
+        -- N.b. It assumes oscoin#550 is implemented.
+        SGetBlocks -> \Range{..} -> timed maxAllowedNetworkLatency $ do
+            let url = stext
```

Comment:
Do we really want to operate with strings here, and parse every time?

Reply:
Uhm, probably not -- perhaps we should pass a function from `Peer -> Request` ? We wouldn't know which port/ip we want to call (that's intrinsic of the `Peer`) but surely the `Request` can be computed beforehand, so that we can avoid the parsing every time.